packages/web3-common/src/web3_base_provider.ts

```diff
@@ -0,0 +1,79 @@
+import {
+	JSONRPCPayload,
+	JSONRPCRequest,
+	JSONRPCResponseWithError,
+	JSONRPCResponseWithResult,
+	JSONRPCResult,
+} from './types';
+
+export interface ProviderMessage<T = JSONRPCResult> {
+	type: string;
+	data: T;
+}
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#connectivity
+export type Web3BaseProviderStatus = 'connecting' | 'connected' | 'disconnected';
+export type Web3BaseProviderCallback<T = JSONRPCResult> = (
+	error: Error | null,
+	result?: ProviderMessage<T>,
+) => void;
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#provider-errors
+export const JSONRPC_ERR_REJECTED_REQUEST = 4001;
+export const JSONRPC_ERR_UNAUTHORIZED = 4001;
+export const JSONRPC_ERR_UNSUPPORTED_METHOD = 4200;
+export const JSONRPC_ERR_DISCONNECTED = 4900;
+export const JSONRPC_ERR_CHAIN_DISCONNECTED = 4901;
+
+// Provider interface compatible with EIP-1193
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md
+export abstract class Web3BaseProvider {
+	// TODO: For legacy support, should be deprecated and removed in favor of `request`
+	public send<T = JSONRPCResult, T2 = unknown[]>(
```

Comment:
And even here where `JSONRPCResult` and `unknown[]` are used multiple times, I believe this to decrease readability

Reply:
All the generics passed down for the extensibility. If user knows exact type while calling `send` he can just pass the type otherwise the default types will be used.